sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c

```diff
@@ -1091,24 +1090,24 @@ static size_t _iphc_ipv6_encode(gnrc_pktsnip_t *pkt,
             }
             gnrc_netif_release(iface);
 
-            if ((ipv6_hdr->src.u64[1].u64 == iid.uint64.u64) ||
+            if ((!memcmp(&ipv6_hdr->src.u8[8], &iid, sizeof(iid))) ||
                 _context_overlaps_iid(src_ctx, &ipv6_hdr->src, &iid)) {
                 /* 0 bits. The address is derived from link-layer address */
                 iphc_hdr[IPHC2_IDX] |= IPHC_SAC_SAM_L2;
                 addr_comp = true;
             }
-            else if ((byteorder_ntohl(ipv6_hdr->src.u32[2]) == 0x000000ff) &&
-                     (byteorder_ntohs(ipv6_hdr->src.u16[6]) == 0xfe00)) {
+            else if (!memcmp(&ipv6_hdr->src.u8[8],
+                             addr_fffe, sizeof(addr_fffe))) {
                 /* 16 bits. The address is derived using 16 bits carried inline */
                 iphc_hdr[IPHC2_IDX] |= IPHC_SAC_SAM_16;
-                memcpy(iphc_hdr + inline_pos, ipv6_hdr->src.u16 + 7, 2);
+                memcpy(iphc_hdr + inline_pos, ipv6_hdr->src.u8 + 14, 2);
```

Comment:
While pointer arithmetics are cool, time taught me, that it is _waaay_ clearer if you use array notation ;-)

```suggestion
                memcpy(iphc_hdr + inline_pos, &ipv6_hdr->src.u8[14], 2);
```